rewrite-core/src/main/java/org/openrewrite/ScanningRecipe.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2023 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite;
+
+import org.openrewrite.internal.lang.Nullable;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.UUID;
+
+/**
+ * A recipe that may first scan a repository and study it in one pass over the
+ * repository's source files before, in another pass, it decides how to transform
+ * the code.
+ * <br/>
+ * New source file generation is part of this type as well, since in almost every case
+ * we check that a file doesn't yet exist (and perhaps other conditions) before deciding
+ * to generate a file.
+ *
+ * @param <T> The type of the accumulator where scanning data is held until the transformation phase.
+ */
+public abstract class ScanningRecipe<T> extends Recipe {
+    private final String recipeAccMessage = "org.openrewrite.recipe.acc." + UUID.randomUUID();
+
+    /**
+     * @return The initial value of the accumulator before any source files have been iterated over.
+     */
+    public abstract T getInitialValue();
```

Comment:
This design forces `T` to be a mutable object, which means that files must always be visited in a single thread.

This constraint could be resolved if `T` were required to be immutable, and this required that `T` either implement a `Mergable` interface, or if there were some other external `SemiGroup` that described how to merge two `T` types together.